### ISSUE TITLE
Missing or uninitialized constant: Bundler::Deprecate when installing a bundle with Rubinius

### ISSUE THREAD

**Issue by philr** [bundler/bundler#6163]
Running `bundle install` with Bundler version 1.16.0 on Rubinius can fail with a 'Missing or uninitialized constant: Bundler::Deprecate' error.

The error only occurs when the Gemfile references gems that are not already installed (or when forcing gems to be installed with `bundle install --force`).

I encounter the error with clean installs of Rubinius versions 3.82 and 3.86 using both RubyGems versions 2.6.14 and 2.7.2.

Running `bundle install --force` with Bundler version 1.15.4 works correctly on Rubinius 3.82 and 3.86.

This issue appears to have been caused by commit f9599c7e8c6775bb4044ad2b4809a957ef8fadea.

## Backtrace

```
NameError: Missing or uninitialized constant: Bundler::Deprecate
  core/module.rb:548:in `const_missing'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/deprecate.rb:22:in `__script__'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/deprecate.rb:10:in `__script__'
  core/code_loader.rb:233:in `require'
  core/kernel.rb:851:in `gem_original_require (require)'
  /home/philr/.rbenv/versions/rbx-3.86/site/rubygems/core_ext/kernel_require.rb:55:in `require'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/psyched_yaml.rb:30:in `__script__'
  core/code_loader.rb:233:in `require'
  core/kernel.rb:851:in `gem_original_require (require)'
  /home/philr/.rbenv/versions/rbx-3.86/site/rubygems/core_ext/kernel_require.rb:55:in `require'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/rubygems_integration.rb:115:in `configuration'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/fetcher.rb:242:in `connection'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/fetcher.rb:85:in `initialize'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/source/rubygems.rb:251:in `fetchers'
  core/array.rb:93:in `map'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/source/rubygems.rb:249:in `fetchers'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/source/rubygems.rb:383:in `remote_specs'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/index.rb:11:in `build'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/source/rubygems.rb:382:in `remote_specs'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/source/rubygems.rb:89:in `specs'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/lazy_specification.rb:76:in `__materialize__'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/spec_set.rb:85:in `materialize'
  core/array.rb:113:in `map!'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/spec_set.rb:82:in `materialize'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/definition.rb:170:in `specs'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/definition.rb:158:in `resolve_remotely!'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/installer.rb:279:in `resolve_if_needed'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/installer.rb:83:in `run'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/process_lock.rb:12:in `lock'
  core/io.rb:1391:in `open'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/process_lock.rb:9:in `lock'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/installer.rb:72:in `run'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/installer.rb:25:in `install'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/cli/install.rb:65:in `run'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/cli.rb:224:in `install'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/settings.rb:136:in `temporary'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/cli.rb:223:in `install'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/cli.rb:27:in `dispatch'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/cli.rb:18:in `start'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/exe/bundle:30:in `__script__'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
  /home/philr/.rbenv/versions/rbx-3.86/gems/gems/bundler-1.16.0/exe/bundle:22:in `__script__'
  core/kernel.rb:572:in `load'
  /home/philr/.rbenv/versions/rbx-3.86/bin/bundle:23:in `__script__'
  core/code_loader.rb:505:in `load_script'
  core/code_loader.rb:590:in `load_script'
  core/loader.rb:679:in `script'
  core/loader.rb:861:in `main'
```

## Environment

```
Bundler       1.16.0
  Platforms   ruby, x86_64-linux, x86_64-rubinius-3.86
Ruby          2.3.1p0 (2017-09-27 revision 0) [x86_64-linux-gnu]
  Full Path   /home/philr/.rbenv/versions/rbx-3.86/bin/rbx
  Config Dir  /home/philr/.rbenv/versions/rbx-3.86/etc
RubyGems      2.6.14
  Gem Home    /home/philr/.rbenv/versions/rbx-3.86/gems
  Gem Path    /home/philr/.rbenv/versions/rbx-3.86/gems:/home/philr/.gem/rbx/2.3
  User Path   /home/philr/.gem/rbx/2.3
  Bin Dir     /home/philr/.rbenv/versions/rbx-3.86/gems/bin
OpenSSL
  Compiled    OpenSSL 1.0.2g  1 Mar 2016
  Loaded      OpenSSL 1.0.2g  1 Mar 2016
  Cert File   /usr/lib/ssl/cert.pem
  Cert Dir    /usr/lib/ssl/certs
Tools
  Git         2.15.0
  RVM         not installed
  rbenv       rbenv 1.1.1-6-g2d7cefe
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2017-10-31
Git SHA           10f20fa33
Released Version  true
```

### Gemfile

```ruby
source "https://rubygems.org"

gem 'tzinfo'
```

### Gemfile.lock

```
GEM
  remote: https://rubygems.org/
  specs:
    thread_safe (0.3.6)
    tzinfo (1.2.4)
      thread_safe (~> 0.1)

PLATFORMS
  ruby

DEPENDENCIES
  tzinfo

BUNDLED WITH
   1.15.4
```

**Comment by akhramov** [comment]
@segiddins  `defined? Bundler::Deprecate` always returns true on Rubinius due to autoloading differences with MRI (rubinius/rubinius#3769)

Also, as bundler uses autoloading to load `Bundler::Deprecate` constant, I'm wondering if it's necessary to require `bundler/deprecate`:
https://github.com/bundler/bundler/blob/f9599c7e8c6775bb4044ad2b4809a957ef8fadea/lib/bundler/psyched_yaml.rb#L30-L37

**Comment by segiddins** [comment]
It’s necessary in case that file is required before “bundler”, which is what declares the autoload

**Comment by mach-kernel** [comment]
I am having this same issue with `rubinius 3.84 (2.3.1 3970f17d 2017-08-02 5.0.0) [x86_64-darwin16.7.0]`

**Comment by brixen** [comment]
The `Bundler::Deprecate` issue is fixed in Rubinius 3.93 (thanks to @akhramov for opening the issue and providing an initial repro). This issue can be close.

There is a concurrency issue with Bundler 1.16.1 that is being tracked here https://github.com/rubinius/rubinius/issues/3781.

**Comment by segiddins** [comment]
> The Bundler::Deprecate issue is fixed in Rubinius 3.93

Thanks @brixen ! I hope to be able to look into #6274 soon, but life has been very hectic recently
